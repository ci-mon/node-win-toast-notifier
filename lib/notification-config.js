// https://learn.microsoft.com/en-us/uwp/schemas/tiles/toastschema/schema-root
export var NotificationLoopingSounds;
(function (NotificationLoopingSounds) {
    NotificationLoopingSounds["Alarm1"] = "ms-winsoundevent:Notification.Looping.Alarm";
    NotificationLoopingSounds["Alarm2"] = "ms-winsoundevent:Notification.Looping.Alarm2";
    NotificationLoopingSounds["Alarm3"] = "ms-winsoundevent:Notification.Looping.Alarm3";
    NotificationLoopingSounds["Alarm4"] = "ms-winsoundevent:Notification.Looping.Alarm4";
    NotificationLoopingSounds["Alarm5"] = "ms-winsoundevent:Notification.Looping.Alarm5";
    NotificationLoopingSounds["Alarm6"] = "ms-winsoundevent:Notification.Looping.Alarm6";
    NotificationLoopingSounds["Alarm7"] = "ms-winsoundevent:Notification.Looping.Alarm7";
    NotificationLoopingSounds["Alarm8"] = "ms-winsoundevent:Notification.Looping.Alarm8";
    NotificationLoopingSounds["Alarm9"] = "ms-winsoundevent:Notification.Looping.Alarm9";
    NotificationLoopingSounds["Alarm10"] = "ms-winsoundevent:Notification.Looping.Alarm10";
    NotificationLoopingSounds["Call1"] = "ms-winsoundevent:Notification.Looping.Call";
    NotificationLoopingSounds["Call2"] = "ms-winsoundevent:Notification.Looping.Call2";
    NotificationLoopingSounds["Call3"] = "ms-winsoundevent:Notification.Looping.Call3";
    NotificationLoopingSounds["Call4"] = "ms-winsoundevent:Notification.Looping.Call4";
    NotificationLoopingSounds["Call5"] = "ms-winsoundevent:Notification.Looping.Call5";
    NotificationLoopingSounds["Call6"] = "ms-winsoundevent:Notification.Looping.Call6";
    NotificationLoopingSounds["Call7"] = "ms-winsoundevent:Notification.Looping.Call7";
    NotificationLoopingSounds["Call8"] = "ms-winsoundevent:Notification.Looping.Call8";
    NotificationLoopingSounds["Call9"] = "ms-winsoundevent:Notification.Looping.Call9";
    NotificationLoopingSounds["Call10"] = "ms-winsoundevent:Notification.Looping.Call10";
})(NotificationLoopingSounds || (NotificationLoopingSounds = {}));
export var NotificationSounds;
(function (NotificationSounds) {
    NotificationSounds["Default"] = "ms-winsoundevent:Notification.Default";
    NotificationSounds["IM"] = "ms-winsoundevent:Notification.IM";
    NotificationSounds["Mail"] = "ms-winsoundevent:Notification.Mail";
    NotificationSounds["Reminder"] = "ms-winsoundevent:Notification.Reminder";
    NotificationSounds["SMS"] = "ms-winsoundevent:Notification.SMS";
})(NotificationSounds || (NotificationSounds = {}));
//# sourceMappingURL=notification-config.js.map