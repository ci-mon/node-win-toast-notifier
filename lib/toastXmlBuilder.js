import { XMLBuilder } from 'fast-xml-parser';
function mapToAttributes(obj, propsToSkip) {
    if (!obj) {
        return {};
    }
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
        if (!value || propsToSkip?.includes(key)) {
            continue;
        }
        result[`@_${key}`] = value.toString();
    }
    return result;
}
function createTextNode(obj, contentProperty, propsToSkip) {
    return {
        '#text': obj[contentProperty],
        ...mapToAttributes(obj, [...propsToSkip, contentProperty]),
    };
}
function buildBody(body) {
    if (typeof body === 'string') {
        return { text: body };
    }
    else {
        const result = {};
        for (const item of body) {
            let items = result[item.type];
            if (!items) {
                items = [];
                result[item.type] = items;
            }
            switch (item.type) {
                case 'text':
                    items.push(createTextNode(item, 'content', ['type']));
                    break;
                case 'image':
                    items.push({
                        ...mapToAttributes(item, ['type']),
                    });
                    break;
            }
        }
        return result;
    }
}
function buildAction(actionConfig) {
    switch (actionConfig.actionType) {
        case 'action':
            return mapToAttributes(actionConfig, ['actionType']);
    }
    return {};
}
function buildInput(input) {
    switch (input.type) {
        case 'text':
            return mapToAttributes(input, ['actionType']);
        case 'selection':
            return {
                ...mapToAttributes(input, ['actionType', 'selection']),
                selection: input.selection.map((s) => mapToAttributes(s)),
            };
    }
    return {};
}
function buildActions(config) {
    if (!config.actions) {
        return {};
    }
    let result = {
        action: [],
        input: [],
    };
    for (const action of config.actions) {
        result[action.actionType].push(action);
    }
    return {
        actions: {
            input: result.input
                .filter((x) => x.actionType === 'input')
                .map(buildInput),
            action: result.action
                .filter((x) => x.actionType === 'action')
                .map(buildAction),
        },
    };
}
export function buildNotificationXml(config) {
    const builder = new XMLBuilder({
        ignoreAttributes: false,
    });
    let jObj = {
        toast: {
            ...(config.audio ? { audio: mapToAttributes(config.audio) } : {}),
            ...(config.header
                ? { header: mapToAttributes(config.header) }
                : {}),
            visual: {
                binding: {
                    '@_template': 'ToastGeneric',
                    ...buildBody(config.body),
                },
            },
            ...buildActions(config),
        },
    };
    return builder.build(jObj);
}
//# sourceMappingURL=toastXmlBuilder.js.map