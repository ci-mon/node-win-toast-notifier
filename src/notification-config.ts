// https://learn.microsoft.com/en-us/uwp/schemas/tiles/toastschema/schema-root

export type NotificationAction = {
    actionType: 'action';
    content: string;
    imageUri?: string;
    arguments: string;
    activationType?: 'foreground' | 'background' | 'protocol';
    placement?: 'contextMenu';
    'hint-inputId'?: string;
    'hint-buttonStyle'?: 'Success' | 'Critical';
    'hint-toolTip'?: string;
};

export type NotificationActionInput = {
    actionType: 'input';
    id: string;
};

export type NotificationActionTextInput = NotificationActionInput & {
    type: 'text';
    placeHolderContent?: string;
};

export type NotificationActionSelectionInput = NotificationActionInput & {
    type: 'selection';
    defaultInput?: string;
    selection: NotificationActionSelectionItem[];
};

export type NotificationActionSelectionItem = {
    id: string;
    content: string;
};

export type NotificationInput =
    | NotificationActionTextInput
    | NotificationActionSelectionInput;

export enum NotificationLoopingSounds {
    Alarm1 = 'ms-winsoundevent:Notification.Looping.Alarm',
    Alarm2 = 'ms-winsoundevent:Notification.Looping.Alarm2',
    Alarm3 = 'ms-winsoundevent:Notification.Looping.Alarm3',
    Alarm4 = 'ms-winsoundevent:Notification.Looping.Alarm4',
    Alarm5 = 'ms-winsoundevent:Notification.Looping.Alarm5',
    Alarm6 = 'ms-winsoundevent:Notification.Looping.Alarm6',
    Alarm7 = 'ms-winsoundevent:Notification.Looping.Alarm7',
    Alarm8 = 'ms-winsoundevent:Notification.Looping.Alarm8',
    Alarm9 = 'ms-winsoundevent:Notification.Looping.Alarm9',
    Alarm10 = 'ms-winsoundevent:Notification.Looping.Alarm10',
    Call1 = 'ms-winsoundevent:Notification.Looping.Call',
    Call2 = 'ms-winsoundevent:Notification.Looping.Call2',
    Call3 = 'ms-winsoundevent:Notification.Looping.Call3',
    Call4 = 'ms-winsoundevent:Notification.Looping.Call4',
    Call5 = 'ms-winsoundevent:Notification.Looping.Call5',
    Call6 = 'ms-winsoundevent:Notification.Looping.Call6',
    Call7 = 'ms-winsoundevent:Notification.Looping.Call7',
    Call8 = 'ms-winsoundevent:Notification.Looping.Call8',
    Call9 = 'ms-winsoundevent:Notification.Looping.Call9',
    Call10 = 'ms-winsoundevent:Notification.Looping.Call10',
}

export enum NotificationSounds {
    Default = 'ms-winsoundevent:Notification.Default',
    IM = 'ms-winsoundevent:Notification.IM',
    Mail = 'ms-winsoundevent:Notification.Mail',
    Reminder = 'ms-winsoundevent:Notification.Reminder',
    SMS = 'ms-winsoundevent:Notification.SMS',
}

export type NotificationAudioConfig = {
    silent?: boolean;
    loop?: boolean;
    src?: NotificationSounds | NotificationLoopingSounds | string;
};

export type NotificationSettings = {
    image_cache?: {
        enable?: true,
        expiration_timeout_seconds?: number
    }
}
export type NotificationConfig = {
    body: string | NotificationBody;
    actions?: Array<NotificationAction | NotificationInput>;
    audio?: NotificationAudioConfig;
    header?: NotificationHeader;
    settings?: NotificationSettings
};

export type NumericId = {
    id?: number;
};

export type NotificationImage = NumericId & {
    type: 'image';
    src: string;
    alt?: string;
    addImageQuery?: boolean;
    placement?: 'appLogoOverride' | 'hero';
    'hint-crop'?: 'circle';
};

export type NotificationText = NumericId & {
    type: 'text';
    lang?: string;
    placement?: 'attribution';
    content: string;
};

export type NotificationBody = Array<NotificationImage | NotificationText>;

export type NotificationHeader = {
    id: string;
    title: string;
    arguments: string;
    activationType?: 'foreground' | 'protocol';
};
