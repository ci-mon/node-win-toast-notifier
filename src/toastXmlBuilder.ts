import {
    NotificationAction,
    NotificationAudioConfig,
    NotificationBody,
    NotificationConfig,
    NotificationInput,
} from './notification-config.js';
import { XMLBuilder } from 'fast-xml-parser';

type KeyOf<T extends object> = Extract<keyof T, string>;

function mapToAttributes<T extends Record<string, unknown>>(
    obj?: T,
    propsToSkip?: Array<KeyOf<T>>
): Record<string, string> {
    if (!obj) {
        return {};
    }
    const result: Record<string, string> = {};
    for (const [key, value] of Object.entries(obj)) {
        if (!value || (propsToSkip as string[])?.includes(key)) {
            continue;
        }
        result[`@_${key}`] = value.toString();
    }
    return result;
}

function createTextNode<T extends Record<string, unknown>>(
    obj: T,
    contentProperty: KeyOf<T>,
    propsToSkip: Array<KeyOf<T>>
): Record<string, string> {
    return {
        '#text': obj[contentProperty] as string,
        ...mapToAttributes(obj, [...propsToSkip, contentProperty]),
    };
}

function buildBody(body: string | NotificationBody) {
    if (typeof body === 'string') {
        return { text: body };
    } else {
        const result: Record<string, unknown> = {};
        for (const item of body) {
            let items = result[item.type] as Array<unknown>;
            if (!items) {
                items = [];
                result[item.type] = items;
            }
            switch (item.type) {
                case 'text':
                    items.push(createTextNode(item, 'content', ['type']));
                    break;
                case 'image':
                    items.push({
                        ...mapToAttributes(item, ['type']),
                    });
                    break;
            }
        }
        return result;
    }
}

function buildAction(actionConfig: NotificationAction) {
    switch (actionConfig.actionType) {
        case 'action':
            return mapToAttributes(actionConfig, ['actionType']);
    }
    return {};
}

function buildInput(input: NotificationInput) {
    switch (input.type) {
        case 'text':
            return mapToAttributes(input, ['actionType']);
        case 'selection':
            return {
                ...mapToAttributes(input, ['actionType', 'selection']),
                selection: input.selection.map((s) => mapToAttributes(s)),
            };
    }
    return {};
}

function buildActions(config: NotificationConfig) {
    if (!config.actions) {
        return {};
    }
    let result: { action: NotificationAction[]; input: NotificationInput[] } = {
        action: [],
        input: [],
    };
    for (const action of config.actions) {
        result[action.actionType].push(action as any);
    }
    return {
        actions: {
            input: result.input
                .filter((x) => x.actionType === 'input')
                .map(buildInput),
            action: result.action
                .filter((x) => x.actionType === 'action')
                .map(buildAction),
        },
    };
}

export function buildNotificationXml(config: NotificationConfig): string {
    const builder = new XMLBuilder({
        ignoreAttributes: false,
    });
    let jObj = {
        toast: {
            ...(config.audio ? { audio: mapToAttributes(config.audio) } : {}),
            ...(config.header
                ? { header: mapToAttributes(config.header) }
                : {}),
            visual: {
                binding: {
                    '@_template': 'ToastGeneric',
                    ...buildBody(config.body),
                },
            },
            ...buildActions(config),
        },
    };
    return builder.build(jObj);
}
