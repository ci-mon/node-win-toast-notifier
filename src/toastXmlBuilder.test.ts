import { test } from '@jest/globals';
import { buildNotificationXml } from './toastXmlBuilder.js';
import { NotificationSounds } from './notification-config.js';

describe('builder tests', () => {
    test('when body is string', () => {
        const result = buildNotificationXml({
            body: 'hello',
        });
        expect(result).toMatchSnapshot();
    });
    test('when body is object', () => {
        const result = buildNotificationXml({
            body: [
                {
                    type: 'text',
                    content: 'Test header',
                },
                {
                    type: 'text',
                    content: 'Test attribution',
                    placement: 'attribution',
                },
                {
                    type: 'image',
                    src: 'my-img',
                    placement: 'appLogoOverride',
                    'hint-crop': 'circle',
                    alt: 'Logo',
                },
                {
                    type: 'image',
                    src: 'my-img-2',
                },
            ],
        });
        expect(result).toMatchSnapshot();
    });
    test('when actions specified', () => {
        const result = buildNotificationXml({
            body: 'Hello',
            actions: [
                {
                    actionType: 'action',
                    content: 'Some button',
                    arguments: 'someButtonClicked',
                },
                {
                    actionType: 'input',
                    type: 'text',
                    placeHolderContent: 'Reply',
                    id: 'replyId',
                },
                {
                    actionType: 'action',
                    content: 'Send',
                    arguments: 'sendReply',
                    'hint-inputId': 'replyId',
                },
                {
                    actionType: 'input',
                    type: 'selection',
                    id: 'vars',
                    defaultInput: 'yep',
                    selection: [
                        {
                            id: 'yep',
                            content: 'Yes',
                        },
                        {
                            id: 'no',
                            content: 'Nope',
                        },
                    ],
                },
            ],
        });
        expect(result).toMatchSnapshot();
    });
    test('when audio specified', () => {
        const result = buildNotificationXml({
            body: 'Hello',
            audio: {
                src: NotificationSounds.SMS,
                silent: false,
                loop: false,
            },
        });
        expect(result).toMatchSnapshot();
    });
    test('when header specified', () => {
        const result = buildNotificationXml({
            body: 'Hello',
            header: {
                id: 'test',
                title: 'some title',
                arguments: 'def args',
            },
        });
        expect(result).toMatchSnapshot();
    });
});
