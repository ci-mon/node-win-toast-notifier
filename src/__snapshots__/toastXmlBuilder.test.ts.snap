// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder tests when actions specified 1`] = `
<toast>
    <visual>
        <binding template="ToastGeneric">
            <text>
                Hello
            </text>
        </binding>
    </visual>
    <actions>
        <input type="text" placeHolderContent="Reply" id="replyId"></input>
        <input type="selection" id="vars" defaultInput="yep">
            <selection id="yep" content="Yes"></selection>
            <selection id="no" content="Nope"></selection>
        </input>
        <action content="Some button" arguments="someButtonClicked"></action>
        <action content="Send" arguments="sendReply" hint-inputId="replyId"></action>
    </actions>
</toast>
`;

exports[`builder tests when audio specified 1`] = `
<toast>
    <audio src="ms-winsoundevent:Notification.SMS"></audio>
    <visual>
        <binding template="ToastGeneric">
            <text>
                Hello
            </text>
        </binding>
    </visual>
</toast>
`;

exports[`builder tests when body is object 1`] = `
<toast>
    <visual>
        <binding template="ToastGeneric">
            <text>
                Test header
            </text>
            <text placement="attribution">
                Test attribution
            </text>
            <image src="my-img" placement="appLogoOverride" hint-crop="circle" alt="Logo"></image>
            <image src="my-img-2"></image>
        </binding>
    </visual>
</toast>
`;

exports[`builder tests when body is string 1`] = `
<toast>
    <visual>
        <binding template="ToastGeneric">
            <text>
                hello
            </text>
        </binding>
    </visual>
</toast>
`;

exports[`builder tests when header specified 1`] = `
<toast>
    <header id="test" title="some title" arguments="def args"></header>
    <visual>
        <binding template="ToastGeneric">
            <text>
                Hello
            </text>
        </binding>
    </visual>
</toast>
`;
